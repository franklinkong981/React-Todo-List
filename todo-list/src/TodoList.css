.TodoList-title {
  color: green;
}

.TodoList-todos {
  width: 80vw;
  padding: 20px;
  border: 2px solid red;
  border-radius: 50px;
  background: linear-gradient(#f06df2, #f7bf47);
}