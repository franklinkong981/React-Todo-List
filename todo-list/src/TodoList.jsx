import react, {useState} from "react";
import {v4 as uuid} from "uuid";

import NewTodoForm from "./NewTodoForm.jsx";
import Todo from "./Todo.jsx";

const TodoList = () => {
  //each todo in the todos state array will be an object containing the id, content (string describing what to do) and whether the todo has been completed or not.
  const [todos, setTodos] = useState([]);

  //once the form is submitted, a new todo is added, it will be given a random uuid as well as start off as not completed.
  const addTodo = (todo) => {
    setTodos(todos => {
      return [...todos, {id: uuid(), content: todo, completed: false}];
    });
  };

  //remove the todo with the specific id from the todos state array.
  const removeTodo = (todoId) => {
    setTodos(todos => {
      return todos.filter((todo) => todo.id !== todoId);
    });
  }

  //if the specific todo is completed, make it not completed, and vice versa.
  const toggleTodo = (todoId) => {
    setTodos(todos => {
      return todos.map(todo => {
        return todo.id == todoId ? todo : {...todo, completed: !(todo.completed)};
      });
    });
  }

  return (
    <div className="TodoList">
      <NewTodoForm addTodoFunction={addTodo} />
      <h2 className="TodoList-title">Todo List:</h2>
      <div className="TodoList-todos">
        {todos.map(todo => <Todo key={todo.id} id={todo.id} content={todo.content} completed={todo.completed} 
        removeTodoFunction={removeTodo} toggleTodoFunction={toggleTodo}/>)}
      </div>
    </div>
  );
};

export default TodoList;