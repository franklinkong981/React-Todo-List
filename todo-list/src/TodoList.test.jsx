import React from "react";
import {render, fireEvent} from "@testing-library/react";
import TodoList from "./TodoList.jsx";
import { expect, it } from "vitest";

it("renders without crashing", function() {
  render(<TodoList/>);
});

it("matches snapshot", function() {
  const {asFragment} = render(<TodoList/>);
  expect(asFragment()).toMatchSnapshot();
});

it("should add a new todo to the page after the form is submitted AND the user should be able to toggle the todo and remove it", async function() {
  const {getByLabelText, queryByText, findByText} = render(<TodoList/>);

  //at first, there shouldn't be any todos on the page.
  let todoToggleButton = queryByText("Mark as complete");
  expect(todoToggleButton).not.toBeInTheDocument();
  let todoRemoveButton = queryByText("Delete Todo");
  expect(todoRemoveButton).not.toBeInTheDocument();
  //the form, however, should be present.
  let formSubmitButton = queryByText("Add New Todo");
  expect(formSubmitButton).toBeInTheDocument();

  //the user should be able to change the form's input value.
  const todoInput = getByLabelText("Enter New Todo:");
  expect(todoInput.value).toEqual("");
  fireEvent.change(todoInput, {target: {value: "Pass all the tests"}});
  expect(todoInput.value).toEqual("Pass all the tests");

  //The user should be able to submit the form and a new todo corresponding to the form's input should be added to the page.
  fireEvent.click(formSubmitButton);
  todoToggleButton = await findByText("Mark as complete");
  expect(todoToggleButton).toBeInTheDocument();
  todoRemoveButton = await findByText("Delete Todo");
  expect(todoRemoveButton).toBeInTheDocument();

  //The form's todo input field should be reset back to blank.
  expect(todoInput.value).toEqual("");

  //The user should be able to mark the todo as complete and unmark it as incomplete.
  let todoText = todoToggleButton.previousSibling;
  expect(todoText).toHaveTextContent("Pass all the tests");
  expect(todoText).toHaveStyle({textDecoration: "none"});
  fireEvent.click(todoToggleButton);
  expect(todoText).toHaveStyle({textDecoration: "line-through"});
  expect(todoToggleButton).toHaveTextContent("Unmark as incomplete");
  fireEvent.click(todoToggleButton);
  expect(todoText).toHaveStyle({textDecoration: "none"});
  expect(todoToggleButton).toHaveTextContent("Mark as complete");

  //The user should be able to delete the todo and it should no longer be on the page.
  fireEvent.click(todoRemoveButton);
  expect(todoToggleButton).not.toBeInTheDocument();
  expect(todoRemoveButton).not.toBeInTheDocument();
});