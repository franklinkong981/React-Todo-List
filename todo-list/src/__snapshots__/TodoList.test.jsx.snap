// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <form
      class="NewTodoForm"
    >
      <label
        class="NewTodoForm-input-label"
        for="todo-input-field"
      >
        Enter New Todo:
      </label>
      <input
        class="NewTodoForm-todo-input"
        id="todo-input-field"
        name="content"
        size="50"
        type="text"
        value=""
      />
      <button
        class="NewTodoForm-submit-button"
        type="submit"
      >
        Add New Todo
      </button>
    </form>
    <h2
      class="TodoList-title"
    >
      Todo List:
    </h2>
    <div
      class="TodoList-todos"
    />
  </div>
</DocumentFragment>
`;
